src/conn.c

```diff
@@ -33,30 +34,114 @@ typedef struct conn_mode_entry {
+     MUTEX_INITIALIZER, NULL, 0, 0},
     {NULL, NULL, conn_thread_init, conn_thread_cleanup, conn_thread_lock, conn_thread_unlock,
-     conn_thread_interrupt, conn_thread_send, conn_thread_get_addrs, MUTEX_INITIALIZER, NULL}};
+     conn_thread_interrupt, conn_thread_send, conn_thread_get_addrs, MUTEX_INITIALIZER, NULL, 0, 0}};
 
 static conn_mode_entry_t *get_mode_entry(juice_agent_t *agent) {
 	juice_concurrency_mode_t mode = agent->config.concurrency_mode;
 	assert(mode >= 0 && mode < MODE_ENTRIES_SIZE);
 	return mode_entries + (int)mode;
 }
 
+// accepts a host name and a port and returns a token that can be used to
+// distinguish one mux request handler from another. E.g.
+//
+// '127.0.0.1', 8080 -> '127.0.0.1:8080'
+// '::', 8080 -> '[::]:8080'
+// NULL, 8080 -> 'any:8080'
+// '', 8080 -> 'any:8080'
+static char *get_address (const char *bind_address, uint16_t port) {
+	if (!bind_address || strcmp(bind_address, "") == 0) {
+		bind_address = "any";
+	}
+
+	// search for '.' in bind_address, treat as IPv4 if found
+	char *result = strchr(bind_address, '.');
+	int index = (int)(result - bind_address);
+	int maxAddrSize = 48; // ip6 is 39 chars + [ + ] + : + 5 for the port + \0
+	char *address = (char*) calloc(1, maxAddrSize * sizeof(char));
+
+	// ip6
+	char *format = "[%s]:%d";
+
+	if (index > -1) {
+		// ip4
+		format = "%s:%d";
+	}
+
+	sprintf(address, format, bind_address, port);
+
+	return address;
+}
+
+static conn_registry_t *get_port_registry(conn_mode_entry_t *entry, const char *bind_address, uint16_t port) {
+	char *address = get_address(bind_address, port);
+
+	for (int i = 0; i < entry->registries_size; i++) {
+		if (!entry->registries[i]) {
+			continue;
+		}
+
+		if (strcmp(entry->registries[i]->address, address) == 0) {
+			return entry->registries[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int add_registry(conn_mode_entry_t *entry, conn_registry_t *registry) {
+	int i = 0;
+		while (i < entry->registries_size && entry->registries[i])
+			++i;
+
+	if (i == entry->registries_size) {
+		int new_size = entry->registries_size * 2;
+
+		if (new_size == 0) {
+			new_size = 1;
+		}
+
+		JLOG_DEBUG("Reallocating registries array, new_size=%d", new_size);
+		assert(new_size > 0);
+
+		conn_registry_t **new_registries =
+				realloc(entry->registries, new_size * sizeof(conn_registry_t *));
+		if (!new_registries) {
+			JLOG_FATAL("Memory reallocation failed for registries array");
+			return -1;
+		}
+
+		entry->registries = new_registries;
+		entry->registries_size = new_size;
+		memset(entry->registries + i, 0, (new_size - i) * sizeof(conn_registry_t *));
+	}
+
+	entry->registries[i] = registry;
+	registry->registry_index = i;
+	++entry->registries_count;
+
+	return 0;
+}
+
 static int acquire_registry(conn_mode_entry_t *entry, udp_socket_config_t *config) {
 	// entry must be locked
-	conn_registry_t *registry = entry->registry;
+	conn_registry_t *registry = get_port_registry(entry, config->bind_address, config->port_begin);
```

Comment:
Here you should only perform the lookup for mux entries, otherwise get the first one.

Reply:
This is done now